python/submit.py

```diff
@@ -18,20 +19,24 @@
 import numpy
 
 
-def parse_time(time):
-    try:
-        time = datetime.datetime.fromisoformat(time)
-
-        if time.tzinfo:
-            # Ensure time is converted to UTC, then drop tzinfo
-            time = time.astimezone(datetime.timezone.utc)
-            time = time.replace(tzinfo=None)
-
-        return time
-    except AttributeError:
-        # Workaround for Python version <3.7, which doesn't have
-        # fromisoformat()
-        return datetime.datetime.strptime(time, '%Y-%m-%dT%H:%M:%S')
+def parse_timestamp(s):
+    if sys.version_info < (3, 11):
+        # in Python 3.10, fromisoformat() does not accept
+        # a trailing Z, and it can only parse millisecond
+        # resolution. We remove any trailing Z, and parse
+        # the fractional seconds separately
+        s = s.rstrip('Z')
+        s_int, s_frac = s.split('.')
+        t_int = datetime.datetime.fromisoformat(s_int)
+        t_frac = float(f'0.{s_frac}')
+        # Truncate t_frac to integer microseconds to match
+        # behaviour of fromisoformat() in newer Python version
+        t = t_int + datetime.timedelta(
+            microseconds=int(t_frac * 1000000))
+    else:
+        t = datetime.datetime.fromisoformat(s)
+    t = t.replace(tzinfo=datetime.timezone.utc)
+    return t
```

Comment:
I have done more testing and spotted an obvious bug in the Python < 3.11 implementation: the timestamp may not have a fractional part!

This is a possible fix:
```suggestion
def parse_timestamp(s):
    t = None
    if sys.version_info < (3, 11):
        # in Python 3.10, fromisoformat() does not accept a trailing Z, and it
        # can only parse millisecond resolution. If there is fractional part, we
        # remove any trailing Z, and parse the fractional seconds separately
        s = s.rstrip('Z')
        if '.' in s:
            s_split = s.split('.')
            s_int, s_frac = s.split('.')
            t_int = datetime.datetime.fromisoformat(s_int)
            t_frac = float(f'0.{s_frac}')
            # Truncate t_frac to integer microseconds to match behaviour of
            # fromisoformat() in newer Python version
            t = t_int + datetime.timedelta(
                microseconds=int(t_frac * 1000000))
    if t is None:
        t = datetime.datetime.fromisoformat(s)
    t = t.replace(tzinfo=datetime.timezone.utc)
    return t
```